azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -70,6 +72,17 @@ def decrypt_p7m(self, p7m_file, trans_prv_file, trans_cert_file, pem_file):
         rc = shellutil.run(cmd)
         if rc != 0:
             logger.error("Failed to decrypt {0}".format(p7m_file))
+    
+    def decrypt_encrypted_file(self, privateKey, encryptedFile):
+        cmd = "{0} cms -decrypt -inform DER -inkey {1} -in {2}".format(self.openssl_cmd, privateKey, encryptedFile)
```

Comment:
I tried echoing the value in and command substitution as well as writing to stdin, but with binary data, I don't see how to do this without a temp file.  I'd love to do this without it, but I don't see a way.

Reply:
You have the base64-encrypted string. Decode that via `binary_blob = base64.b64decode(base64str)`. Then, instead of using `shellutil.run_get_output()`, you'll have to use Python's subprocess module directly. You can most easily use subprocess.Popen() and .communicate() to send `binary_blob` as the stdin of the command.

Meanwhile, you'd set the command to `{0} cms -decrypt -inform DER -inkey {1} -in /dev/stdin".format(etc...)`. This tells the cms command that the data to be decrypted is arriving on stdout.

See https://stackoverflow.com/questions/163542/python-how-do-i-pass-a-string-into-subprocess-popen-using-the-stdin-argument for an example of how to do this.

Reply:
Thank you.  I will try to eliminate this with my next iteration

---
In reply to: [192201846](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192201846) [](ancestors = 192201846)